src/server/routes/entity/entity.tsx

```diff
@@ -84,67 +84,68 @@ export function displayEntity(req: PassportRequest, res: $Response) {
-			bbid: resLocals.entity.bbid,
-			editorId: resLocals.user.id
-		})
-			.save(null, {method: 'insert'})
-			.then(() => achievement.processPageVisit(orm, resLocals.user.id))
-			.catch(
-				// error caused by duplicates we do not want in database
-				() => new Promise(resolve => resolve(false))
-			);
+		try {
+			await new EditorEntityVisits({
+				bbid: resLocals.entity.bbid,
+				editorId: resLocals.user.id
+			})
+				.save(null, {method: 'insert'});
+
+			editorEntityVisitPromise = await achievement.processPageVisit(orm, resLocals.user.id);
+		}
+		catch {
+			// error caused by duplicates we do not want in database
+			return false;
+		}
 	}
 	else {
 		editorEntityVisitPromise = new Promise(resolve => resolve(false));
 	}
 
-	let alertPromise = editorEntityVisitPromise.then((visitAlert) => {
-		let alertIds = [];
-		if (visitAlert.alert) {
-			alertIds = alertIds.concat(visitAlert.alert.split(',').map(
-				(id) => parseInt(id, 10)
-			));
-		}
-		if (_.isString(req.query.alert)) {
-			// $FlowFixMe
-			alertIds = alertIds.concat(req.query.alert.split(',').map(
-				(id) =>	parseInt(id, 10)
-			));
-		}
-		if (alertIds.length > 0) {
-			const promiseList = alertIds.map(
-				(achievementAlert) =>
-					new AchievementUnlock(
-						{id: achievementAlert}
-					)
-						.fetch({
-							require: true,
-							withRelated: 'achievement'
-						})
-						.then((unlock) => unlock.toJSON())
-						.then((unlock) => {
-							let unlockName;
-							if (req.user.id === unlock.editorId) {
-								unlockName = {
-									name: unlock.achievement.name
-								};
-							}
-							return unlockName;
-						})
-						.catch((err) => {
-							log.debug(err);
-						})
-			);
-			alertPromise = Promise.all(promiseList);
-		}
-		else {
-			alertPromise = new Promise(resolve => resolve(false));
-		}
-		return alertPromise;
-	});
+	let alertPromise;
+	const visitAlert = await editorEntityVisitPromise;
+	let alertIds = [];
+	if (visitAlert.alert) {
+		alertIds = alertIds.concat(visitAlert.alert.split(',').map(
+			(id) => parseInt(id, 10)
+		));
+	}
+	if (_.isString(req.query.alert)) {
+		// $FlowFixMe
+		alertIds = alertIds.concat(req.query.alert.split(',').map(
+			(id) =>	parseInt(id, 10)
+		));
+	}
+	if (alertIds.length > 0) {
+		const promiseList = alertIds.map(async (achievementAlert) => {
+			try {
+				const unlock = await new AchievementUnlock({id: achievementAlert}).fetch({
+					require: true,
+					withRelated: 'achievement'
+				});
+				const unlockJSON = unlock.toJSON();
 
-	return alertPromise.then((alert) => {
+				let unlockName;
+				if (req.user.id === unlockJSON.editorId) {
+					unlockName = {
+						name: unlockJSON.achievement.name
+					};
+				}
+				return unlockName;
+			}
+			catch (err) {
+				return log.debug(err);
+			}
+		});
+
+		alertPromise = Promise.all(promiseList);
+	}
+	else {
+		alertPromise = new Promise(resolve => resolve(false));
+	}
+
+	return (async () => {
+		const alert = await alertPromise;
```

Comment:
In this particular case, `alertPromise` is (almost) well named since we do store `Promise.all` or `new Promise` *without awaiting them*.

I would still rename it to `alertPromises` since the main reason we have it is so we cant `await`  multiple promises with `Promise.all`.

Reply:
why do we have to do alertPromise = Promise.all(...)  when we already awaited the promises inside map function
```
const unlock = await new AchievementUnlock({id: achievementAlert}).fetch({
					require: true,
					withRelated: 'achievement'
				});
```
unless we want to catch individual promise rejection we may need not to use promise.all() here ........

Reply:
same question here
```
entityJSON.relationshipSet.relationships = await Promise.all(entityJSON.relationshipSet.relationships.map(async (rel) => {
				try {
					rel.source = await commonUtils.getEntityAlias(orm, rel.source.bbid, rel.source.type);
					rel.target = await commonUtils.getEntityAlias(orm, rel.target.bbid, rel.target.type);
				}
				catch (err) {
					log.error(err);
				}
				return rel;
```

Reply:
You're right, there's some subtleties here.

In short, array.map takes an array and runs the callback function on each item, then takes the return value. The resulting array would be composed of each return value.
In our case, the return value is a Promise, not the value returned from the callback function. That's why we would call `await Promise.all` to get the actual values.

Here's an example with an array of three numbers, and a callback function that awaits a timeout and returns the value.
You'll see the first arrayMapResult is an array of Promises, while the second using `await Promise.all(…` is an array of the actual return values.

<img width="452" alt="image" src="https://github.com/metabrainz/bookbrainz-site/assets/6179856/9d991133-83be-49dc-933e-8733df153010">

So in the first case, that .map work like a .forEach and we can't do anything with the returned array, while we would use Promise.all if we want to do something with the returned values.
In both cases, the timeout is awaited.
Hope that clarifies it a bit!

I'll let you think through what we should be returning in which case, but do ping me if you're struggling.

Reply:
thanks a bunch ! Concept clear 💯💯